esmvalcore/local.py

```diff
@@ -448,14 +448,115 @@ def find_files(self, **facets) -> list[LocalFile]:
+          filename_template: "{f3}[._]{f4}*"
+          --> regex_pattern:
+          "/root/(?P<f2>[^_/]*?)/(?P<f3>[^_/]*?)[\._](?P<f4>[^_/]*?).*?"
+        - rootpath: "/root"
+          dirname_template: "{f1}/{f1}-{f2}"
+          filename_template: "*.nc"
+          --> regex_pattern:
+          "/root/(?P<f1>[^_/]*?)/(?P=f1)\-(?P<f2>[^_/]*?)/.*?\.nc"
+        - rootpath: "/root"
+          dirname_template: "{f1}/{f2}{f3}"
+          filename_template: "*.nc"
+          --> regex_pattern:
+          "/root/(?P<f1>[^_/]*?)/(?:[^_/]*?)/.*?\.nc"
+
+        """
+        dirname_template = self.dirname_template
+        filename_template = self.filename_template
+
+        # Templates must not be absolute paths (i.e., start with /), otherwise
+        # the roopath is ignored (see
+        # https://docs.python.org/3/library/pathlib.html#operators)
+        if self.dirname_template.startswith(os.sep):
+            dirname_template = dirname_template[1:]
+        if self.filename_template.startswith(os.sep):
+            filename_template = filename_template[1:]
+
+        pattern = re.escape(
+            str(self.rootpath / dirname_template / filename_template)
+        )
+
+        # Remove all tags that are in between other tags, e.g.,
+        # {tag1}{tag2}{tag3} -> {tag1}{tag2} (there is no way to reliably
+        # extract facets from those)
+        pattern = re.sub(r"(?<=\})\\\{[^\}]+?\\\}(?=\\(?=\{))", "", pattern)
+
+        # Replace consecutive tags, e.g. {tag1}{tag2} with non-capturing groups
+        # (?:[^_/]*?) (there is no way to reliably extract facets from those)
+        # Note: This assumes that facets do NOT contain / or _
+        pattern = re.sub(
+            r"\\\{[^\{]+?\}\\\{[^\}]+?\\\}", rf"(?:[^_{os.sep}]*?)", pattern
+        )
+
+        # Convert tags {tag} to named capture groups (?P<tag>[^_/]*?); for
+        # duplicates use named backreferences (?P=tag)
+        # Note: This assumes that facets do NOT contain / or _
+        already_used_tags: set[str] = set()
+        for full_tag in re.findall(r"\\\{(.+?)\\\}", pattern):
+            # Ignore .upper and .lower (full_tag: {tag.lower}, tag: {tag})
+            if full_tag.endswith(r"\.upper") or full_tag.endswith(r"\.lower"):
+                tag = full_tag[:-7]
+            else:
+                tag = full_tag
+
+            old_str = rf"\{{{full_tag}\}}"
+            if tag in already_used_tags:
+                new_str = rf"(?P={tag})"
+            else:
+                new_str = rf"(?P<{tag}>[^_{os.sep}]*?)"
+                already_used_tags.add(tag)
+
+            pattern = pattern.replace(old_str, new_str, 1)
+
+        # Convert fnmatch wildcards * and [] to regex wildcards
+        pattern = pattern.replace(r"\*", ".*?")
+        for chars in re.findall(r"\\\[(.*?)\\\]", pattern):
+            pattern = pattern.replace(rf"\[{chars}\]", f"[{chars}]")
+
+        return pattern
+
+    @property
+    def regex_pattern(self) -> str:
```

Comment:
It would be nicer to use a [`__post_init__`](https://docs.python.org/3/library/dataclasses.html#dataclasses.__post_init__) method to initialize the `regex_pattern` attribute. Would that be possible?